mlflow/data/dataset_source.py

```diff
@@ -0,0 +1,92 @@
+import json
+
+from abc import abstractmethod
+from typing import TypeVar, Any, Dict
+
+
+DatasetSourceType = TypeVar("DatasetSourceType", bound="DatasetSource")
+
+
+class DatasetSource:
+    """
+    Represents the source of a dataset used in MLflow Tracking, providing information such as
+    cloud storage location, delta table name / version, etc.
+    """
+
+    @staticmethod
+    @abstractmethod
+    def _get_source_type() -> str:
+        """
+        Obtains a string representing the source type of the dataset.
+
+        :return: A string representing the source type of the dataset, e.g. "s3", "delta_table", ...
+        """
+
+    @abstractmethod
+    def download(self) -> Any:
+        """
+        Downloads the files / objects referred to be this DatasetSource.
+
+        :return: The downloaded source, e.g. a local filesystem path, a Spark DataFrame, etc.
+        """
+
+    @staticmethod
+    @abstractmethod
+    def _can_resolve(raw_source: Any) -> bool:
```

Comment:
Just a thought -- would it make more sense to collapse `_can_resolve` with `_resolve` into something like `_resolve_if_possible`?